### ISSUE TITLE
Support app group container with synchronized Realm

### ISSUE THREAD

**Issue by ewerx** [realm/realm-cocoa#4490]
## Goals

I am currently sharing a realm db between an iOS app and its today widget extension by storing the realm in the application group container directory and setting this path in the Realm configuration's `fileURL`:

```
var config = Realm.Configuration()
config.fileURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupIdentifier)!.appendingPathComponent(dbFilename)
Realm.Configuration.defaultConfiguration = config
```

Now I want to add cloud sync features to the app using Realm Mobile Platform, but it is not possible to specify both a `fileURL` and a `syncConfiguration` in the Configuration. How can I enable synchronization and have both the app and extension continue to access the existing db that's stored in the group container?

Related StackOverflow question (not mine): https://stackoverflow.com/questions/40785346/swift-realm-mobile-platform-today-extension

## Expected Results

It's possible to synchronize a Realm db stored in the app group container directory.

**Comment by austinzheng** [comment]
Thanks for getting in touch with us!

If you plan to have some synced database X being used by both the app extension and the main app, you can have each one open a separate copy and sync will keep them up to date. I don't think sync currently supports migrating a non-synced Realm directly to a synced Realm, although all of this will be changing in the near future.

Hope that helps!

**Comment by ewerx** [comment]
So if I understand correctly, both app extension and the app should specify a `SyncConfiguration` to access the remote db instead of using the app group container, and each will maintain their own separate local cache of the db for offline access?

One disadvantage to this approach is that the app extension requires internet access to remain synchronized with the parent app. If the device is offline the widget/extension will remain out of sync until a connection is re-established.

That's not a showstopper in my case, but if there currently no way to migrate existing local Realm db data to a synced Realm I will have to postpone deploying cloud sync features or look at other options like implementing sync via CloudKit on my own.

**Comment by austinzheng** [comment]
Yes, that's the situation as it currently stands. We're definitely interested in ways to make the use case you described easier to implement, though, and I'll pass your feedback along. Thanks!

**Comment by ewerx** [comment]
Thanks. It would definitely be useful to be able to sync a realm and store the local db in the app group container for extension access. If this is not already on the roadmap I would like to make a feature request to keep track of its status.

**Comment by austinzheng** [comment]
Thanks for your feedback. I'll add it to our backlog.

**Comment by brock190** [comment]
Are there any further updates on this issue?

I am sharing the local db with the today widget and Watch, so it would be good if it was possible to 
synchronize a Realm db stored in the app group container directory,
even if migrating a non-synced Realm directly to a synced Realm was not supported.

**Comment by austinzheng** [comment]
This issue is still in our backlog because, for technical and business reasons, we've prioritized other issues ahead of it. If that changes we'll update this thread.

**Comment by mommyme** [comment]
Wow, I went through almost completely converting my app to Realm and then I saw this. Back to Core Data for me. Can't justify not using two separate databases for app and app extension. Bummer. So disappointed right now.

**Comment by jpsim** [comment]
Core Data also doesn't support inter-process access unless you're willing to forgo fine-grained notifications. Although support for that appears to be coming in iOS 11.

**Comment by xcodewarrier** [comment]
As an addition data point, we're also removing Realm from our app because of this issue.

**Comment by xcodewarrier** [comment]
Out of interest it was claimed that this was solved back in 2015, what has changed?
 https://news.realm.io/news/thomas-goyne-fast-inter-process-communication/

**Comment by brock190** [comment]
We are in the same boat having implemented synced realms, but no support for app group container directory is a showstopper as we cannot justify having a separate system for the app extension.

Are there any plans to support this in the near future?

**Comment by austinzheng** [comment]
The article linked above deals with non-synchronized Realms, which work differently from synchronized Realms in a number of ways that are usually not relevant to end users.

We have a very large list of things we'd like to add support for, but our sync engineers simply cannot make any guarantees at this time as to when this feature will be available. We regret the inconvenience.

**Comment by ewerx** [comment]
I was under the impression that apps with synchronized Realms use a local copy (whose location is not user-configurable) for all operations , and the sync logic happens behind the scenes (thus allowing seamless offline access to the db).

If there was an option to store this local copy in the app group container, would that be enough to allow extensions to access the main app's Realm directly without requiring their own synchronized Realm configuration and separate local copy? 

Even restricting extensions to read-only access would still be useful, since extensions could request the parent app to make state changes.

**Comment by xcodewarrier** [comment]
+1 @ewerx restricting extensions to read-only access would be more than sufficient for our needs

**Comment by austinzheng** [comment]
The problem isn't figuring out where to store the file. The underlying problem is that the synchronization subsystem only works correctly if at most one instance of the sync client has the file open at any given time, and there's currently no way for Realm to coordinate between a parent app and extension process to ensure that this is true.

**Comment by gerbil802** [comment]
If the Backlog label was replaced with an Enhancement label I wonder if it means it's on the roadmap =) The last comment made it sound almost unfeasible.

**Comment by Alex293** [comment]
Any update on this ?

**Comment by chuckIT94** [comment]
To reduce the file size now - is it handy to use partialSync for this ? For a share extension which wants to write to the database thats also not very handy with a synched realm.

**Comment by Marlon-Monroy** [comment]
any updates on this?

**Comment by rouuuge** [comment]
does anybody has a workaround to use siri with sync realm?

**Comment by dianaafanador3** [comment]
Hi @rouuuge Can you give explain a little bit in detail your use case?. If it is not related to supporting a synchronised realm in app group container, please create a new issue.

**Comment by rouuuge** [comment]
sorry just could solved it. I had to login on my app intent with the credentials I shared from the mainapp. so my use case is solved.

**Comment by almas-ims** [comment]
It's been 5 years, any hope on this?

**Comment by jsflax** [comment]
We've a project scheduled to work on this in the coming quarters, so, there is hope.

**Comment by nick-condon** [comment]
Any update on this? I’m just about to remove Realm completely from my app because updating Realm via the share extension while is a core function I want to add.
